File: Perez-Meneses-Juan-Carlos/Pipes/pipes.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>

typedef struct {
    int fd[2];        // Descriptores de archivo para la tubería
    const char* buf;  // Datos a escribir en la tubería
} PipeWriter;

// Función para inicializar el PipeWriter
void initPipeWriter(PipeWriter* writer, const char* buffer) {
    writer->buf = buffer;
    if (pipe(writer->fd) < 0) {
        perror("Error en la creación del pipe");
        exit(EXIT_FAILURE);
    }
}

// Función para escribir en la tubería
void writeToPipe(PipeWriter* writer) {
    int i = 0;
    while (1) {
        write(writer->fd[1], writer->buf, 1); // Escribe 1 byte en la tubería
        printf("%d ", i++);
        fflush(stdout); // Asegura que el buffer de salida se imprima
    }
}

// Función para liberar recursos
void closePipeWriter(PipeWriter* writer) {
    close(writer->fd[0]);
    close(writer->fd[1]);
}

int main() {
    const char* buffer = "a";
    PipeWriter writer;

    initPipeWriter(&writer, buffer);
    writeToPipe(&writer);
    closePipeWriter(&writer);

    return 0;
}

File: Perez-Meneses-Juan-Carlos/Pipes/pipes.m
#import <Foundation/Foundation.h>
#include <unistd.h>
#include <stdio.h>

@interface PipeWriter : NSObject {
    int fd[2]; // Descriptores de archivo para la tubería
    const char *buf;
}

- (instancetype)initWithBuffer:(const char *)buffer;
- (void)writeToPipe;
- (void)closePipe;

@end

@implementation PipeWriter

- (instancetype)initWithBuffer:(const char *)buffer {
    self = [super init];
    if (self) {
        buf = buffer;
        if (pipe(fd) < 0) {
            NSLog(@"Error en la creación del pipe");
            exit(1);
        }
    }
    return self;
}

- (void)writeToPipe {
    int i = 0;
    while (1) {
        write(fd[1], buf, 1); // Escribe un byte en la tubería
        printf("%d ", i++);
        fflush(stdout); // Asegura que el buffer de salida se imprima
    }
}

- (void)closePipe {
    close(fd[0]);
    close(fd[1]);
}

@end

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        const char *buffer = "a";
        PipeWriter *writer = [[PipeWriter alloc] initWithBuffer:buffer];
        [writer writeToPipe];
        [writer closePipe];
    }
    return 0;
}

